### PR TITLE
feat: Adjusted how routemaps work

### PR DESCRIPTION
**Pull request by nickrandolph**
GitHub Issue (If applicable): #

## PR Type

What kind of change does this PR introduce?

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Tested code with current [supported SDKs](../README.md#supported)
- [ ] Docs have been added/updated which fit [documentation template](https://github.com/nventive/Uno/blob/master/doc/.feature-template.md). (for bug fixes / features)
- [ ] [Unit Tests and/or UI Tests](doc/articles/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] [Wasm UI Tests](doc/articles/working-with-the-samples-apps.md#running-the-webassembly-ui-tests-snapshots) are not showing unexpected any differences. Validate PR `Screenshots Compare Test Run` results.
- [ ] Contains **NO** breaking changes
- [ ] Updated the [Release Notes](https://github.com/nventive/Uno/tree/master/doc/ReleaseNotes)
- [ ] Associated with an issue (GitHub or internal)

## Other information

Internal Issue (If applicable):

### CODE REVIEW

File: src/Uno.Extensions.Navigation.Toolkit/Uno.Extensions.Navigation.Toolkit.UI.csproj [hunk 1]
```diff
@@ -19,4 +19,20 @@
 	</ItemGroup>
 	
 	<Import Project="common.props" />
+	
+	<Choose>
+		<When Condition="'$(TargetFramework)'=='xamarinios10' or '$(TargetFramework)'=='monoandroid10.0' or '$(TargetFramework)'=='monoandroid11.0'">
```
Comment (nickrandolph) [inline src/Uno.Extensions.Navigation.Toolkit/Uno.Extensions.Navigation.Toolkit.UI.csproj:24]: @kazo0 this needs to include the net6-android and net6-ios tfms - I can add these later today

**Comment by nickrandolph** [comment]
@jeromelaban think I'm going to need your help on resolving the outstanding build issue. It seems to be somehow related to the navigation.ui/winui projects but for the life of me I can't work out why Droid (WinUI) is failing to build. It seems to be the only project where this is an issue

File: samples/Commerce/Commerce.Shared/Views/CartPage.xaml [hunk 2]
```diff
@@ -47,7 +47,9 @@
 											   Style="{StaticResource MaterialSubtitle2}" />
 								</utu:AutoLayout>
 								<ListView ItemsSource="{Binding Items}"
-										  uen:Navigation.Request="ProductDetails"
+										  uen:Navigation.Request=""
```
Comment (jeromelaban) [inline samples/Commerce/Commerce.Shared/Views/CartPage.xaml:50]: What does it mean to navigate to an empty string ?

File: samples/Commerce/WinUI/Commerce.Wasm/LinkerConfig.xml [hunk 3]
```diff
@@ -1,7 +1,35 @@
...
+	<assembly fullname="Serilog.Sinks.Console" />
+	<assembly fullname="Serilog.Sinks.Debug" />
+	<assembly fullname="Serilog.Sinks.File" />
+	<assembly fullname="Microsoft.Extensions.Options" />
+	<assembly fullname="Microsoft.Extensions.Hosting" />
+	<assembly fullname="Microsoft.Extensions.Logging" />
+	<assembly fullname="Microsoft.Extensions.Logging.Configuration" />
+	<assembly fullname="Microsoft.Extensions.Logging.Console" />
+	<assembly fullname="Microsoft.Extensions.Logging.Debug" />
+	<assembly fullname="Microsoft.Extensions.Logging.EventSource" />
```
Comment (jeromelaban) [inline samples/Commerce/WinUI/Commerce.Wasm/LinkerConfig.xml:32]: we're going to have to be finer grained that this, otherwise the AOT build is going to be very large.

File: src/Uno.CrossTargeting.props [hunk 4]
```diff
@@ -1,6 +1,6 @@
 <Project ToolsVersion="15.0">
-	<Import Project="$(SolutionDir)\crosstargeting_override.props" Condition="exists('$(SolutionDir)\crosstargeting_override.props')" />
-
+	<!-- <Import Project="$(SolutionDir)\crosstargeting_override.props" Condition="exists('$(SolutionDir)\crosstargeting_override.props')" /> -->
+	<Import Project="crosstargeting_override.props" Condition="exists('crosstargeting_override.props')" />
```
Comment (jeromelaban) [inline src/Uno.CrossTargeting.props:3]: Why did you need this change ?

File: samples/Commerce/Commerce.Shared/ViewModels/ProductDetailsViewModel.cs [hunk 5]
```diff
@@ -20,9 +20,9 @@ public partial class ProductDetailsViewModel
 		_product = product;
 	}
 
-	public IFeed<Product> Product => Feed.Async(Load);
+	public virtual IFeed<Product> Product => Feed.Async(Load);
 
-	public IFeed<Review[]> Reviews => Product.SelectAsync(async (p, ct) => await _productService.GetReviews(p.ProductId, ct));
+	public virtual IFeed<Review[]> Reviews => Product.SelectAsync(async (p, ct) => await _productService.GetReviews(p.ProductId, ct));
```
Comment (jeromelaban) [inline samples/Commerce/Commerce.Shared/ViewModels/ProductDetailsViewModel.cs:25]: @dr1rrb is virtual needed here ?

File: src/Uno.Extensions.Navigation.Toolkit/Navigators/TabBarNavigator.cs [hunk 6]
```diff
@@ -1,29 +1,16 @@
...
-using Microsoft.UI.Xaml;
-using Microsoft.UI.Xaml.Controls;
-#endif
 
 namespace Uno.Extensions.Navigation.Toolkit.Navigators;
 
-public class TabBarNavigator: ControlNavigator<TabBar>
+public class TabBarNavigator : ControlNavigator<TabBar>
 {
+	protected override bool RequiresDefaultView => true;
```
Comment (jeromelaban) [inline src/Uno.Extensions.Navigation.Toolkit/Navigators/TabBarNavigator.cs:12]: Why is a default view required ?

File: src/Uno.Extensions.Navigation.UI/Navigator.cs [hunk 7]
```diff
@@ -110,7 +86,9 @@ private async Task InitializeRootNavigator()
 
 			// Is this region is an unnamed child of a composite,
 			// send request to parent if the route has no scheme
-			if ((request.Route.IsCurrent() || request.Route.IsBackOrCloseNavigation()) &&
+			if ((request.Route.IsCurrent()
+				//|| request.Route.IsBackOrCloseNavigation()
```
Comment (jeromelaban) [inline src/Uno.Extensions.Navigation.UI/Navigator.cs:90]: Is this still needed ?

File: src/Uno.Extensions.Navigation.UI/Navigator.cs [hunk 8]
```diff
@@ -178,7 +158,12 @@ private async Task InitializeRootNavigator()
 
 		var responseFactory = services.GetRequiredService<IResponseNavigatorFactory>();
 		// Create ResponseNavigator if result is requested
-		var navigator = request.GetResponseNavigator(responseFactory, this);
+		var navigator = request.Result is not null ? request.GetResponseNavigator(responseFactory, this) : default;
+
+		//if(navigator is not null)
+		//{
+		//	request = request with { Result = null };
+		//}
```
Comment (jeromelaban) [inline src/Uno.Extensions.Navigation.UI/Navigator.cs:166]: Still needed ?

File: src/Uno.Extensions.Navigation.UI/Navigators/ControlNavigator.cs [hunk 9]
```diff
@@ -50,9 +38,9 @@ public abstract class ControlNavigator<TControl> : ControlNavigator
         }
 
         var route = request.Route;
-        var mapping = Mappings.FindView(route);
-        Logger.LogDebugMessage($"Navigating to path '{route.Base}' with view '{mapping?.ViewType?.Name}'");
-        var executedPath = await Show(route.Base, mapping?.ViewType, route.Data);
+        var mapping = RouteResolver.Find(route);
+        Logger.LogDebugMessage($"Navigating to path '{route.Base}' with view '{mapping?.View?.Name}'");
```
Comment (jeromelaban) [inline src/Uno.Extensions.Navigation.UI/Navigators/ControlNavigator.cs:42]: We'll need to change this to avoid interpolating message for debug messages. The only pattern that currently works (until we get the delegated interpolation) is `if(Logger.IsEnabled(Debug))`.

File: src/Uno.Extensions.Navigation.UI/Navigators/FlyoutNavigator.cs [hunk 10]
```diff
@@ -1,109 +1,116 @@
...
+		if (viewType is not null)
+		{
+			if (!injectedFlyout)
+			{
+				flyout = Activator.CreateInstance(viewType) as Flyout;
+			}
+			else
+			{
+				flyout = services.GetService<Flyout>();
+				//flyout = new UI.Controls.NavigationFlyout();
```
Comment (jeromelaban) [inline src/Uno.Extensions.Navigation.UI/Navigators/FlyoutNavigator.cs:80]: Remove

File: src/Uno.Extensions.Navigation.UI/Navigators/FrameNavigator.cs [hunk 11]
```diff
@@ -125,7 +123,7 @@ protected override bool CanNavigateToRoute(Route route)
 		//await Show(context.Request.Route.Base, context.Mapping?.View, context.Request.Route.Data);
```
Comment (jeromelaban) [inline src/Uno.Extensions.Navigation.UI/Navigators/FrameNavigator.cs:123]: Remove ?

File: src/Uno.Extensions.Navigation.UI/ResponseNavigator.cs [hunk 12]
```diff
@@ -59,13 +61,13 @@ public ResponseNavigator(INavigator internalNavigation, NavigationRequest reques
...
-			typedResponse.Result.ContinueWith(x => ApplyResult(x.Result));
+		//if (navResponse is NavigationResultResponse<TResult> typedResponse &&
+		//	typedResponse.Result is not null)
+		//{
+		//	typedResponse.Result.ContinueWith(x => ApplyResult(x.Result));
 
-			return typedResponse with { Result = ResultCompletion.Task };
-		}
+		//	return typedResponse with { Result = ResultCompletion.Task };
+		//}
```
Comment (jeromelaban) [inline src/Uno.Extensions.Navigation.UI/ResponseNavigator.cs:70]: Remove ?

File: src/Uno.Extensions.Navigation.UI/RouteResolver.cs [hunk 13]
```diff
@@ -0,0 +1,95 @@
+﻿namespace Uno.Extensions.Navigation;
+
+public class RouteResolver : IRouteResolver
+{
+	protected IList<RouteMap> Mappings { get; } = new List<RouteMap>();
+
+	protected ILogger Logger { get; }
+
+	protected RouteResolver(ILogger logger, IRouteRegistry routes)//, IEnumerable<ViewMap> viewMaps)
```
Comment (jeromelaban) [inline src/Uno.Extensions.Navigation.UI/RouteResolver.cs:9]:
```suggestion
	protected RouteResolver(ILogger logger, IRouteRegistry routes)
```

File: src/Uno.Extensions.Navigation.UI/RouteResolver.cs [hunk 14]
```diff
@@ -0,0 +1,95 @@
...
+		if (maps is not null)
+		{
+			Mappings.AddRange(maps.Flatten());
+		}
+	}
+
+	public RouteResolver(
+		ILogger<RouteResolver> logger,
+		IRouteRegistry routes
+		//, IEnumerable<ViewMap> viewMaps
```
Comment (jeromelaban) [inline src/Uno.Extensions.Navigation.UI/RouteResolver.cs:24]:
```suggestion
```

File: src/Uno.Extensions.Navigation.UI/RouteResolver.cs [hunk 15]
```diff
@@ -0,0 +1,95 @@
...
+		) : this((ILogger)logger, routes
+			//, viewMaps
```
Comment (jeromelaban) [inline src/Uno.Extensions.Navigation.UI/RouteResolver.cs:26]:
```suggestion
```

File: src/Uno.Extensions.Navigation.UI/RouteResolverDefault.cs [hunk 16]
```diff
@@ -0,0 +1,163 @@
...
+
+	private RouteMap? DefaultMapping(string? path = null, Type? view = null, Type? viewModel = null)
+	{
+		if (!ReturnImplicitMapping)
+		{
+			Logger.LogDebugMessage("Implicit mapping disabled");
+			return null;
+		}
+
+		Logger.LogWarningMessage($"For better performance (avoid reflection), create mapping for for path '{path}', view '{view?.Name}', view model '{viewModel?.Name}'");
```
Comment (jeromelaban) [inline src/Uno.Extensions.Navigation.UI/RouteResolverDefault.cs:47]: Same logging comment for warnings

File: src/Uno.Extensions.Navigation.UI/ServiceProviderExtensions.cs [hunk 17]
```diff
@@ -1,15 +1,58 @@
...
+
+	public static void AddInstance<T>(this IServiceProvider provider, Type serviceType, Func<T> instanceCreator)
+	{
+		provider.GetRequiredService<IInstanceRepository>().Instances[serviceType] = instanceCreator;
+	}
+
+	public static T AddInstance<T>(this IServiceProvider provider, T instance)
+	{
+#pragma warning disable CS8604 // Possible null reference argument.
+		provider.AddInstance(typeof(T), instance);
```
Comment (jeromelaban) [inline src/Uno.Extensions.Navigation.UI/ServiceProviderExtensions.cs:21]: Why not a `!` ?

File: src/extensions/Directory.Build.props.old [hunk 18]
```diff
@@ -108,12 +108,12 @@
   </PropertyGroup>-->
 
   <ItemGroup Condition="'$(IsWinUI)' != 'true' and '$(IsUWP)' != 'true' and '$(UnoCompatibility)' != 'UWP'">
-    <PackageReference Include="Uno.WinUI" Version="4.0.9" >
+    <PackageReference Include="Uno.WinUI" Version="4.0.11" >
```
Comment (jeromelaban) [inline src/extensions/Directory.Build.props.old:111]: Remove this file entirely ?

File: samples/Commerce/Commerce.Shared/Views/CartPage.xaml (hunk 2 above)
Comment (nickrandolph) [reply samples/Commerce/Commerce.Shared/Views/CartPage.xaml:50]: It means that navigation should be based on the data type of the item selected from the list. In this case, it's a CartItem - so this type is used to look up the route from the table of RouteMaps

File: samples/Commerce/WinUI/Commerce.Wasm/LinkerConfig.xml (hunk 3 above)
Comment (nickrandolph) [reply samples/Commerce/WinUI/Commerce.Wasm/LinkerConfig.xml:32]: yep, the priority was to get it running. we should be able to pair this back a bit to work out what's essential.

File: src/Uno.CrossTargeting.props (hunk 4 above)
Comment (nickrandolph) [reply src/Uno.CrossTargeting.props:3]: For some reason having the $(SolutionDir) prefix wasn't working - you might have some more insights on this and want to reverse it.

File: samples/Commerce/Commerce.Shared/ViewModels/ProductDetailsViewModel.cs (hunk 5 above)
Comment (nickrandolph) [reply samples/Commerce/Commerce.Shared/ViewModels/ProductDetailsViewModel.cs:25]: Oh, good call - @dr1rrb I needed to make these virtual and override them in the CartProductDetailsViewModel in order to get the code gen to work properly

File: src/Uno.Extensions.Navigation.Toolkit/Navigators/TabBarNavigator.cs (hunk 6 above)
Comment (nickrandolph) [reply src/Uno.Extensions.Navigation.Toolkit/Navigators/TabBarNavigator.cs:12]: Removing this now as not required

File: src/Uno.Extensions.Navigation.UI/Navigator.cs (hunk 7 above)
Comment (nickrandolph) [reply src/Uno.Extensions.Navigation.UI/Navigator.cs:90]: No, I'll remove it

File: src/Uno.Extensions.Navigation.UI/Navigator.cs (hunk 8 above)
Comment (nickrandolph) [reply src/Uno.Extensions.Navigation.UI/Navigator.cs:166]: I'll remove

File: src/Uno.Extensions.Navigation.UI/Navigators/FlyoutNavigator.cs (hunk 10 above)
Comment (nickrandolph) [reply src/Uno.Extensions.Navigation.UI/Navigators/FlyoutNavigator.cs:80]: Done

File: src/Uno.Extensions.Navigation.UI/Navigators/FrameNavigator.cs (hunk 11 above)
Comment (nickrandolph) [reply src/Uno.Extensions.Navigation.UI/Navigators/FrameNavigator.cs]: done

File: src/Uno.Extensions.Navigation.UI/RouteResolver.cs (hunk 14 above)
Comment (nickrandolph) [reply src/Uno.Extensions.Navigation.UI/RouteResolver.cs:24]: done

File: src/Uno.Extensions.Navigation.UI/RouteResolver.cs (hunk 15 above)
Comment (nickrandolph) [reply src/Uno.Extensions.Navigation.UI/RouteResolver.cs:26]: done

File: src/Uno.Extensions.Navigation.UI/ServiceProviderExtensions.cs (hunk 17 above)
Comment (nickrandolph) [reply src/Uno.Extensions.Navigation.UI/ServiceProviderExtensions.cs]: changed

File: src/extensions/Directory.Build.props.old (hunk 18 above)
Comment (nickrandolph) [reply src/extensions/Directory.Build.props.old]: done - removed the old app template and example app

File: src/Uno.Extensions.Navigation.UI/ResponseNavigator.cs (hunk 12 above)
Comment (nickrandolph) [reply src/Uno.Extensions.Navigation.UI/ResponseNavigator.cs:70]: done

File: src/Uno.Extensions.Navigation.UI/RouteResolver.cs (hunk 13 above)
Comment (nickrandolph) [reply src/Uno.Extensions.Navigation.UI/RouteResolver.cs:9]: done

File: src/Uno.Extensions.Navigation.UI/Navigator.cs (hunk 7 above)
Comment (nickrandolph) [reply src/Uno.Extensions.Navigation.UI/Navigator.cs]: removed

File: samples/Commerce/Commerce.Shared/ViewModels/ProductDetailsViewModel.cs (hunk 5 above)
Comment (dr1rrb) [reply samples/Commerce/Commerce.Shared/ViewModels/ProductDetailsViewModel.cs:25]: Ok that because you are inheriting one VM from another, the the code gen is not taking that case in consideration yet. That a nice hack to make those properties virtual to work.

Comment (dr1rrb) [reply samples/Commerce/Commerce.Shared/ViewModels/ProductDetailsViewModel.cs:25]: BTW I just created an issue for that https://github.com/unoplatform/uno.extensions/issues/234

**Review by francoistanguay — APPROVED** [review]
